File: Algorithms/1306_Jump_Game_III/Python/Jump_Game_III_Solution_2.py

# Space: O(n)
# Time: O(n)

# DFS Approach

class Solution:
    def canReach(self, arr, start):
        length = len(arr)
        if length == 1 and start == 0: return True

        self.visited = []

        def helper(index):
            if index in self.visited: return False
            if arr[index] == 0: return True
            left_res, right_res = False, False

            self.visited.append(index)
            if 0 <= index + arr[index] < length: right_res = helper(index + arr[index])
            if 0 <= index - arr[index] < length: left_res = helper(index - arr[index])
            self.visited.pop()

            return left_res or right_res

        return helper(start)

File: Algorithms/1306_Jump_Game_III/Java/JumpGameIIISolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    private int[] cache;

    public boolean canReach(int[] arr, int start) {
        if (arr.length == 1) return arr[0] == 0;

        cache = new int[arr.length];
        Arrays.fill(cache, 0);
        return dfs(arr, start);
    }

    private boolean dfs(int[] arr, int index) {
        if (arr[index] == 0) return true;
        if (cache[index] == 1) return false;
        cache[index] = 1;
        int left = index - arr[index];
        int right = index + arr[index];
        if (0 <= left && left < arr.length) {
            boolean leftRes = dfs(arr, left);
            if (leftRes) return leftRes;
        }
        if (0 <= right && right < arr.length) {
            boolean rightRes = dfs(arr, right);
            if (rightRes) return rightRes;
        }
        cache[index] = 0;
        return false;
    }
}




